Public/Src/Cache/ContentStore/Library/Stores/ContentDirectorySnapshot.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using BuildXL.Cache.ContentStore.Hashing;
+
+namespace BuildXL.Cache.ContentStore.Stores
+{
+    /// <summary>
+    /// This class represents an immutable snapshot of a collection of items at an undetermined point in time, used for enumerations of the file
+    /// system. It is made specifically to avoid issues with enumerations of large amounts of files which go over the maximum object size
+    /// restriction.
+    /// </summary>
+    /// <typeparam name="T">Type held inside the snapshot</typeparam>
+    public class ContentDirectorySnapshot<T> : IEnumerable<PayloadFromDisk<T>>
+    {
+        private readonly List<PayloadFromDisk<T>>[] _snapshot;
+
+        /// <nodoc />
+        public long Count { get; } = 0;
+
+        /// <nodoc />
+        public ContentDirectorySnapshot()
+        {
+            _snapshot = InitializeSnapshot();
+        }
+
+        /// <nodoc />
+        public ContentDirectorySnapshot(IEnumerable<PayloadFromDisk<T>> snapshot)
+        {
+            _snapshot = InitializeSnapshot();
+
+            foreach (var payload in snapshot)
+            {
+                var identifier = payload.Hash[0];

```

Comment:
>var [](start = 16, length = 3)

We should assert that this is a byte, if this is not a byte, then we should use an `unchecked` region to let it overlow into the byte.

Reply:
I see now that all is fine, perhaps a comment stating taking the first byte of the hash to round-robin the entries accross multiple lists.

---
In reply to: [285707794](https://github.com/microsoft/BuildXL/pull/329#discussion_r285707794) [](ancestors = 285707794)